src/mantle/testing/expectation/class-expectation.php

```diff
@@ -167,16 +167,14 @@ public function validate() {
 			}
 
 			// Compare the return value of the hook.
-			if ( '_undefined_' !== $this->return_value ) {
+			if ( isset( $this->return_value_callback ) ) {
 				foreach ( $this->record_stop as $record ) {
-					PHPUnit::assertSame(
-						[ $this->return_value ],
-						$record,
+					PHPUnit::assertTrue(
+						call_user_func_array( $this->return_value_callback, $record ),
 						sprintf(
-							'Failed asserting that hook [%s] return value %s is identical to %s.',
+							'Failed asserting that hook\'s [%s] return value %s matches the expected return value.',
```

Comment:
'Failed asserting that hook [%s]\'s return value %s matches the expected return value.',